File: android/app/src/main/java/org/impfai/hermes/engine/Bm25.kt
package org.impfai.hermes.engine

import kotlin.math.ln

/**
 * Okapi BM25 —— 移植自 backend/hermes_shanghan/rag/bm25.py。
 * 常數與公式逐項一致：k1=1.5, b=0.75,
 * idf = ln(1 + (N - df + 0.5) / (df + 0.5)),
 * score += idf * tf * (k1+1) / (tf + k1 * (1 - b + b * dl/avgdl))。
 */
class Bm25Index(
    private val k1: Double = 1.5,
    private val b: Double = 0.75,
) {
    private val docIds = ArrayList<String>()
    private val docLen = ArrayList<Int>()
    private val df = HashMap<String, Int>()
    private val postings = HashMap<String, MutableList<IntArray>>() // (docIdx, tf)
    private var avgdl = 0.0
    private var finalized = false

    fun add(docId: String, text: String) {
        check(!finalized) { "index already finalized" }
        val toks = TextNorm.tokenize(text)
        val idx = docIds.size
        docIds.add(docId)
        docLen.add(toks.size)
        val counts = HashMap<String, Int>()
        for (t in toks) counts[t] = (counts[t] ?: 0) + 1
        for ((t, c) in counts) {
            df[t] = (df[t] ?: 0) + 1
            postings.getOrPut(t) { ArrayList() }.add(intArrayOf(idx, c))
        }
    }

    fun finalizeIndex() {
        avgdl = if (docLen.isEmpty()) 0.0 else docLen.sum().toDouble() / docLen.size
        finalized = true
    }

    fun search(query: String, topK: Int = 10): List<Pair<String, Double>> {
        if (docIds.isEmpty()) return emptyList()
        val qToks = TextNorm.tokenize(query).toSet()
        val n = docIds.size
        val scores = HashMap<Int, Double>()
        for (t in qToks) {
            val plist = postings[t] ?: continue
            val dfT = df[t] ?: continue
            val idf = ln(1 + (n - dfT + 0.5) / (dfT + 0.5))
            for (p in plist) {
                val i = p[0]
                val tf = p[1].toDouble()
                val dl = (docLen[i].takeIf { it > 0 } ?: 1).toDouble()
                val denom = tf + k1 * (1 - b + b * dl / avgdl)
                scores[i] = (scores[i] ?: 0.0) + idf * tf * (k1 + 1) / denom
            }
        }
        // 與 Python 一致：按原始分排序截斷，返回值 round(s, 4)（銀行家舍入）
        // ——下游 LocalClauseStore 消費的是舍入後的 bm（審查發現 #5）
        return scores.entries
            .sortedByDescending { it.value }
            .take(topK)
            .map { docIds[it.key] to roundHalfEven(it.value, 4) }
    }

    companion object {
        fun roundHalfEven(v: Double, scale: Int): Double =
            java.math.BigDecimal(v)
                .setScale(scale, java.math.RoundingMode.HALF_EVEN)
                .toDouble()
    }
}

File: backend/hermes_shanghan/rag/bm25.py
"""Pure-python BM25 (Okapi) over char unigram+bigram tokens.

No external dependencies; fast enough for the full corpus (≤ tens of
thousands of short classical paragraphs).
"""
from __future__ import annotations

import math
from collections import Counter, defaultdict
from typing import Dict, List, Sequence, Tuple

from ..textutil import tokenize


class BM25Index:
    def __init__(self, k1: float = 1.5, b: float = 0.75):
        self.k1 = k1
        self.b = b
        self.doc_ids: List[str] = []
        self.doc_len: List[int] = []
        self.tf: List[Counter] = []
        self.df: Counter = Counter()
        self.avgdl: float = 0.0

    def add(self, doc_id: str, text: str):
        toks = tokenize(text)
        self.doc_ids.append(doc_id)
        self.doc_len.append(len(toks))
        counts = Counter(toks)
        self.tf.append(counts)
        for t in counts:
            self.df[t] += 1

    def finalize(self):
        self.avgdl = (sum(self.doc_len) / len(self.doc_len)) if self.doc_len else 0.0
        # inverted index for speed
        self._postings: Dict[str, List[Tuple[int, int]]] = defaultdict(list)
        for i, counts in enumerate(self.tf):
            for t, c in counts.items():
                self._postings[t].append((i, c))

    def search(self, query: str, top_k: int = 10) -> List[Tuple[str, float]]:
        if not self.doc_ids:
            return []
        q_toks = set(tokenize(query))
        n_docs = len(self.doc_ids)
        scores: Dict[int, float] = defaultdict(float)
        for t in q_toks:
            postings = self._postings.get(t)
            if not postings:
                continue
            idf = math.log(1 + (n_docs - self.df[t] + 0.5) / (self.df[t] + 0.5))
            for i, tf in postings:
                dl = self.doc_len[i] or 1
                denom = tf + self.k1 * (1 - self.b + self.b * dl / self.avgdl)
                scores[i] += idf * tf * (self.k1 + 1) / denom
        ranked = sorted(scores.items(), key=lambda kv: -kv[1])[:top_k]
        return [(self.doc_ids[i], round(s, 4)) for i, s in ranked]
